src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
This issue is not about `reordering`.  I agree that a memory barrier can avoid `reordering` optimization, and the `pthread_mutex_lock ` contains such barrier.
Here is an example of this `register caching` issue I am actually concerning.
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0472c/CJAEGDEA.html

I also checked the assembly code of this example.
```
int cond = 1;
while(cond) {
    pthread_mutex_lock();
    pthread_mutex_unlock();
    printf("%d", cond);
}
```
The compiler (just tried GCC) does not do the `register caching` or `replacing with constant` for `cond`, is not because of the barrier thing, it is actually because it cannot understand the behavior of external functions such as `pthread_mutex_lock`. I tried to use just `extern void dummy();` to replace these functions, it does the same thing -- the `cond` is reloaded from memory for `printf` line.

Although the current progress engine code coincidently calls external functions, thus should not fall into the issue I am concerning. I am still standing on the side of using `volatile` in the RMA code, just to be safe in case one changes the progress engine code  in future.